### PR TITLE
[AOTInductor] Allow user to explicitly specify Device to run on

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Summary:
AOTInductor currently infer cuda device index by `cudaGetDevice()`. This assumes outer runtime calls `cudaSetDevice()` somewhere, before invoking AOTInductor run. 

This diff adds an explicit argument for specifying target Device. e.g. compiled on "cuda:0", run on "cuda:1". 

todo:
- Are the changes in interface.h BC breaking? as it changes the function signatures in .so file. Might just need introduce a new "Create" function.

Test Plan: CI

Differential Revision:
D52747132

Privacy Context Container: 368960445142440

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

**Comment by chenyang78** [comment]
Please fix the lint failure. Thanks.

File: torch/csrc/inductor/aoti_runtime/interface.h [hunk 1]
```diff
@@ -52,6 +54,15 @@ AOTIRuntimeError AOTInductorModelContainerCreate(
     bool is_cpu,
     const char* cubin_dir);
 
+// Creates an AOTInductor model container. The parameter num_models
+// specifies the number of model instances that may be run concurrently for
+// the same input model.
+AOTIRuntimeError AOTInductorModelContainerCreateWithDevice(
+    AOTInductorModelContainerHandle* container_handle,
+    size_t num_models,
+    const char* device_str,
```
Comment (desertfire) [inline torch/csrc/inductor/aoti_runtime/interface.h:63]: What happens if `device_str == nullptr`? Let's document the behavior by adding comments.

File: torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp [hunk 2]
```diff
@@ -7,7 +7,7 @@ AOTIModelContainerRunnerCuda::AOTIModelContainerRunnerCuda(
     const std::string& model_so_path,
```
Comment (desertfire) [inline torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp:7]: Should we also update the signature of `AOTIModelContainerRunnerCuda`?

**Review by desertfire — APPROVED** [review]

File: torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp (hunk 2 above)
Comment (SherlockNoMad) [reply torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp:7]: done.

File: torch/csrc/inductor/aoti_runtime/interface.h (hunk 1 above)
Comment (SherlockNoMad) [reply torch/csrc/inductor/aoti_runtime/interface.h:63]: commented. 
device_str cannot be nullptr, and it must be a valid device str.

**Review by khabinov — APPROVED** [review]

**Pull request closed without merging**